servant-server/src/Servant/Server/Internal.hs

```diff
@@ -232,6 +259,75 @@ instance
 
   route Proxy = methodRouterHeaders methodDelete (Proxy :: Proxy ctypes) ok200
 
+-- | Authentication in Strict mode.
+instance
+#if MIN_VERSION_base(4,8,0)
+         {-# OVERLAPPABLE #-}
+#endif
+         (AuthData authdata , HasServer sublayout) => HasServer (AuthProtect authdata (usr :: *) 'Strict :> sublayout) where
+
+    type ServerT (AuthProtect authdata usr 'Strict :> sublayout) m = AuthProtected authdata usr (usr -> ServerT sublayout m) 'Strict
+
+    route _ subserver = WithRequest $ \req ->
+        route (Proxy :: Proxy sublayout) $ do
+            -- Note: this may perform IO for each attempt at matching.
+            rr <- routeResult <$> subserver
+
+            case rr of
+                -- Successful route match, so we extract the author-provided
+                -- auth data.
+                Right authProtectionStrict ->
```

Comment:
@jkarni there are two failures here represented by the Nothing:

1. request is missing authentication (e.g.. doesn't have a header)
2. request has authentication, but authentication fails (e.g. wrong password)

in each case there is a different `Response`. So, even though the failure handling is the same (`failWith (RouteMismatch resp)`) the `resp` is different in each case.

I could wrap it in a `EitherT e IO` with a custom `e` or we can rejig the `AuthProtected` data family instance to use `EitherT e IO` as opposed to returning values like `IO (Maybe a)`.